### PR TITLE
Extensive usage of StorageTransmitters

### LINKED ISSUE #1952: Use StorageTransmitters when running an ert3 experiment
Issue spawned by the discussion in #1944 

Scope:

- [ ] Use StorageTransmitters to load data into storage when using the `ert3 records load` command 
- [ ] Use StorageTransmitters when preparing storage records 
- [ ] Use StorageTransmitters when evaluating the experiment/ setting up the prefect config.

### PR DESCRIPTION
**Pull request by DanSava**
**Issue**
Resolves #1952

### CODE REVIEW

File: ert_shared/ensemble_evaluator/ensemble/prefect.py [hunk 1]
```diff
@@ -220,16 +220,16 @@ def context_method():
...
             target=self._evaluate,
             args=(config, ee_id),
         )
+
+        # mp_ctx = multiprocessing.get_context(method=PrefectEnsemble.context_method())
+        # self._eval_proc = mp_ctx.Process(
+        #     target=self._evaluate,
+        #     args=(config, ee_id),
+        # )
         self._eval_proc.daemon = True
```
Comment (sondreso) [inline ert_shared/ensemble_evaluator/ensemble/prefect.py]: Is there a reason why are we changing this? 🤔

Comment (DanSava) [reply ert_shared/ensemble_evaluator/ensemble/prefect.py]: Currently, the change is just tentative we might need to make it permanent if we want to do any kind of posting data from the prefect workers to storage using the async client `httpx`.

But currently, it requires a bit more testing and investigation

File: ert3/engine/_run.py [hunk 2]
```diff
@@ -377,101 +241,122 @@ def _analyze_sensitivity(
...
+        ensemble,
         experiment_config,
         parameters_config,
         workspace_root,
         experiment_name,
+        output_transmitters,
     )
+
+
+def estimate_ensemble_size(
```
[the remark this answers is not in the document]
Comment (DanSava) [reply ert3/engine/_run.py:353]: Not really an estimate, should probably be renamed to something like `get_ensemble_size`

File: ert3/algorithms/_sensitivity.py [hunk 3]
```diff
@@ -1,11 +1,12 @@
+import asyncio
 from typing import Set, MutableMapping, List, Dict, Optional, Any
 
 import numpy as np
 from SALib.sample import fast_sampler
 from SALib.analyze import fast
 
 from ert3.stats import Distribution, Gaussian, Uniform
-from ert.data import RecordCollectionMap, Record, NumericalRecord
+from ert.data import RecordTransmitter, Record, NumericalRecord
```
[the remark this answers is not in the document]
Comment (DanSava) [reply ert3/algorithms/_sensitivity.py:9]: More cleanup is still required but RecordCollectionMap might not be needed anymore

File: ert3/algorithms/_sensitivity.py [hunk 4]
```diff
@@ -131,22 +132,32 @@ def fast_analyze(
...
+    futures = []
+    records = []
+    for iens, transmitter_map in model_output.items():
+        if len(transmitter_map) > 1:
+            raise ValueError("Cannot analyze sensitivity with multiple outputs")
+        if len(transmitter_map) < 1:
+            raise ValueError("Cannot analyze sensitivity with no output")
+        for _, transmitter in transmitter_map.items():
+            futures.append(transmitter.load())
+        if iens > 0 and iens % 50 == 0:
```
[the remark this answers is not in the document]
Comment (DanSava) [reply ert3/algorithms/_sensitivity.py:144]: Yes, it will probably end up as a constant somewhere if I cannot get rid of it.

[the remark this answers is not in the document]
Comment (DanSava) [reply ert3/algorithms/_sensitivity.py:144]: I got rid of it now we do one realization at a time

File: ert/storage/_storage.py [hunk 5]
```diff
@@ -64,27 +63,45 @@ def token(cls) -> str:
...
+        return self._record_type
+
+    def set_transmitted(self, uri: str, record_type: ert.data.RecordType) -> None:
+        self._set_transmitted_state(uri, record_type)
 
     async def _transmit_numerical_record(self, record: ert.data.NumericalRecord) -> str:
         url = f"{self._uri}/matrix"
         if self._real_id is not None:
             url = f"{url}?realization_index={self._real_id}"
+            self._uri = f"{self._uri}?realization_index={self._real_id}"
```
[the remark this answers is not in the document]
Comment (DanSava) [reply ert/storage/_storage.py:90]: Remembered what the issue was: If we only set `self._uri`only in the constructor, we end up with the URL -> `url = f"{self._uri}/matrix"` -> `storage_url/record_name?realization_index=#/matrix` which is not valid

File: ert/storage/_storage.py [hunk 6]
```diff
@@ -99,28 +110,109 @@ async def _load_blob_record(self) -> ert.data.BlobRecord:
...
+
+
 async def _get_from_server_async(
     url: str,
     headers: Dict[str, str],
     **kwargs: Any,
-) -> httpx.Response:
-    async with httpx.AsyncClient() as session:
-        resp = await session.get(url=url, headers=headers, timeout=None, **kwargs)
+) -> requests.Response:
```
Comment (verveerpj) [inline ert/storage/_storage.py:156]: Was it decided to not use httpx now?  It is still in the set.py and release notes, but apparently not used anywhere.

Comment (DanSava) [reply ert/storage/_storage.py:156]: I decided not to use it anymore due to the intermittent errors that appear when dealing with large realizations mainly in the ert3 polynomial example test.

I will remove it from the setup.py also

**Review by verveerpj — COMMENTED** [review]
I don't really have more comments. There are a lot of changes, which are sometimes hard to track, but it seems clear that the code gets simpler in many places, which I like.

I would say it is good, but maybe it is good if someone else has a look, since it is such a large PR.

File: ert/storage/_storage.py [hunk 7]
```diff
@@ -71,20 +69,33 @@ def __init__(self, name: str, storage_url: str, iens: Optional[int] = None):
         self._name: str = name
         self._uri = f"{storage_url}/{name}"
         self._real_id: Optional[int] = iens
-        if self._real_id is not None:
-            self._uri = f"{self._uri}?realization_index={self._real_id}"
+
+    @property
+    def uri(self) -> str:
+        assert self.is_transmitted()
```
Comment (jondequinor) [inline ert/storage/_storage.py:75]:
```suggestion
        if not self.is_transmitted():
            raise RuntimeError(f"Record {self._name} not transmitted")
```
for consistency?

File: ert/storage/_storage.py [hunk 8]
```diff
@@ -212,6 +279,82 @@ async def load_record(url: str, record_type: ert.data.RecordType) -> ert.data.Re
     return ert.data.BlobRecord(data=content)
 
 
+async def get_record_metadata(record_url: str) -> Dict[Any, Any]:
+    headers = {
+        "Token": StorageInfo.token(),
+    }
+    url = f"{record_url}/userdata?realization_index=0"
```
[the remark this answers is not in the document]
Comment (jondequinor) [reply ert/storage/_storage.py:313]: @markusdregi in https://github.com/equinor/ert/issues/1710 you discuss the issue Peter highlights. Is this a continuation of https://github.com/equinor/ert/pull/1693 and thus okay?

File: ert/storage/_storage.py [hunk 9]
```diff
@@ -71,20 +69,33 @@ def __init__(self, name: str, storage_url: str, iens: Optional[int] = None):
...
+        assert self.is_transmitted()
+        return self._uri
+
+    @property
+    def record_type(self) -> Optional[ert.data.RecordType]:
+        assert self.is_transmitted()
+        return self._record_type
+
+    def set_transmitted(self, uri: str, record_type: ert.data.RecordType) -> None:
+        self._set_transmitted_state(uri, record_type)
```
Comment (jondequinor) [inline ert/storage/_storage.py:85]: I'd really like to hide this `_set_transmitted_state` and rather implement a `RecordTransmitter.from_transmitted()` which is general and to be implemented for all transmitter types. Because if this is introduced as-is, the `RecordTransmitter` interface is obsolete.

File: ert/storage/_storage.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jondequinor) [reply ert/storage/_storage.py:91]: What's the underlying cause of this?

File: ert/storage/_storage.py [hunk 10]
```diff
@@ -99,28 +110,109 @@ async def _load_blob_record(self) -> ert.data.BlobRecord:
         return ert.data.BlobRecord(data=record.data)
 
 
+async def get_record_storage_transmitters(
```
Comment (jondequinor) [inline ert/storage/_storage.py:114]: 1. Could this get some documentation?
2. Is this how you get a uniform record, i.e. multiple records given an `ensemble_size` and `record_source`? Why isn't this implemented for all transmitters?

File: .github/workflows/build.yml [hunk 11]
```diff
@@ -155,6 +155,8 @@ jobs:
 
     - name: Test MacOS
       if: matrix.os == 'macos-latest'
+      env:
+        OBJC_DISABLE_INITIALIZE_FORK_SAFETY: 'YES'
```
Comment (jondequinor) [inline .github/workflows/build.yml:159]: Could you comment in-code on why this was added?

File: ert3/engine/_run.py [hunk 12]
```diff
@@ -103,104 +39,50 @@ def _prepare_experiment_record(
     workspace_root: pathlib.Path,
     parameters_config: ert3.config.ParametersConfig,
     experiment_config: ert3.config.ExperimentConfig,
-) -> None:
+) -> Dict[int, Dict[str, ert.storage.StorageRecordTransmitter]]:
+    assert len(record_source) == 2
     if record_source[0] == "storage":
```
Comment (jondequinor) [inline ert3/engine/_run.py:44]: A bit confusing that this is guarded with `storage` whereas a bit down in the PR, the guard is testing against `ert_storage`. The two comparees probably stem from different sources, but should the namespacing of the records use `ert_storage`, not simply `storage`? Maybe I'm misunderstanding something here… 🤔

File: .github/workflows/run_examples_polynomial.yml [hunk 13]
```diff
@@ -44,6 +44,8 @@ jobs:
         ert3 service check storage --timeout 100
 
     - name: Run polynomial demo
+      env:
+        OBJC_DISABLE_INITIALIZE_FORK_SAFETY: 'YES'
```
Comment (jondequinor) [inline .github/workflows/run_examples_polynomial.yml:48]: Same as above

File: tests/ert_tests/ert3/storage/test_storage.py [hunk 14]
```diff
@@ -134,117 +135,52 @@ def test_add_and_get_ensemble_record(tmpdir, raw_ensrec, ert_storage):
...
-@pytest.mark.parametrize(
-    "raw_ensrec",
-    (
-        [{"data": [i + 0.5, i + 1.1, i + 2.2]} for i in range(3)],
-        [{"data": {"a": i + 0.5, "b": i + 1.1, "c": i + 2.2}} for i in range(5)],
-        [{"data": {2: i + 0.5, 5: i + 1.1, 7: i + 2.2}} for i in range(2)],
-        [{"data": b"asdfkasjdhjflkjah21WE123TTDSG34f"}],
-    ),
-)
-def test_add_and_get_ensemble_parameter_record(tmpdir, raw_ensrec, ert_storage):
```
Comment (jondequinor) [inline tests/ert_tests/ert3/storage/test_storage.py:160]: Was this moved or deleted? If the latter, could you comment as to why?

File: ert/storage/_storage.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (DanSava) [reply ert/storage/_storage.py:91]: Because `add_record` expects a fully resolved url

File: ert/storage/_storage.py (hunk 10 above)
Comment (DanSava) [reply ert/storage/_storage.py:114]: > Could this get some documentation? 

Sure what would do you feel needs more documentation?

> Is this how you get a uniform record, i.e. multiple records given an ensemble_size and record_source? 

Yes that is the main idea

> Why isn't this implemented for all transmitters? 

Why would it be useful to make file transmitters or in memory transmitters from records that are already in storage ?

File: ert/storage/_storage.py (hunk 9 above)
Comment (DanSava) [reply ert/storage/_storage.py:85]: Good point!

File: ert/storage/_storage.py (hunk 10 above)
Comment (jondequinor) [reply ert/storage/_storage.py:114]: > Why would it be useful to make file transmitters or in memory transmitters from records that are already in storage ?

Let me rephrase: Is uniform records only supported for ert-storage?

File: ert/storage/_storage.py (hunk 9 above)
Comment (DanSava) [reply ert/storage/_storage.py:85]: After thinking about it a bit more I realized I don't really understand what you would want you would want `RecordTransmitter.from_transmitted()` to return. A general RecordTransmitter ?  Then the implementations for each of the transmitter types would return an instance of that Transmitter type that is filled in with name, uri and record type and the state is set to  `RecordTransmitterState.transmitted`. But if that is the case would it not make more sense to just move the implementation of `set_transmitted` into the `RecordTrasmitter` interface.  

Also, I am not sure how are you thinking the usage would go for an "already" transmitted inmemory transmitter. Not really sure how to wrap my brain around that.

File: ert/storage/_storage.py (hunk 10 above)
Comment (DanSava) [reply ert/storage/_storage.py:114]: In my world view "Uniform records" are a special type of `EnsembleRecords` which currently are only supported for ert-storage so I guess yes they are only supported for ert-storage

File: tests/ert_tests/ert3/storage/test_storage.py (hunk 14 above)
Comment (DanSava) [reply tests/ert_tests/ert3/storage/test_storage.py:160]: I was removed because it was testing the old way `RecordCollections` we added to storage, which has fundamentally changed and is covered by other tests

File: ert3/engine/_run.py (hunk 12 above)
Comment (DanSava) [reply ert3/engine/_run.py:44]: They could both use `storage` for the guard if the ensemble storage type `str` is renamed from `ert_storage` to just `storage`. But currently, they are different to highlight that in one case we are referring to a record source and in another, we are referring to what type of transmitters should be used when "transmitting" data. 

But I agree they could be the same.

File: .github/workflows/build.yml (hunk 11 above)
Comment (DanSava) [reply .github/workflows/build.yml:159]: I will add a comment but I am sure it is not the right fix for the issue that required adding this line. 

If the line is removed. On macos the following error appears:
```
objc[67570]: +[__NSPlaceholderDate initialize] may have been in progress in another thread when fork() was called.
objc[67570]: +[__NSPlaceholderDate initialize] may have been in progress in another thread when fork() was called. We cannot safely call it or ignore it in the fork() child process. Crashing instead. Set a breakpoint on objc_initializeAfterForkError to debug.
```

For reference, this is where I found the fix for the error above: https://stackoverflow.com/questions/50168647/multiprocessing-causes-python-to-crash-and-gives-an-error-may-have-been-in-progr

I could not find the underlying issue that was causing the error and I found it a bit strange because we are using `frokserver` when starting the process that spins up perfect.  

Also, the fun part is that the error appears intermittently on macOS :)

File: ert/storage/_storage.py (hunk 9 above)
Comment (jondequinor) [reply ert/storage/_storage.py:85]: Mainly I wanted to keep the `RecordTransmitter` interface relevant. If the interface is turbulent (which they often are), it may be indicative of a problematic abstraction.

Also, I wanted to protect the "transmitted" state. I see it as a side-effect from doing real work (like actually transmitting stuff), and not really something that should be generally available to mutate.

Does that make sense?

File: ert/storage/_storage.py (hunk 10 above)
Comment (jondequinor) [reply ert/storage/_storage.py:114]: So if I want to use uniform shared-disk records, they currently work (the records themselves are copied _n_ times), but they aren't implemented the same way as the ert-storage records?

File: .github/workflows/build.yml (hunk 11 above)
Comment (jondequinor) [reply .github/workflows/build.yml:161]: Seems clear from the SO answer that you're laxing security introduced in High Sierra?

So maybe the comment could be:

`# Relax multithreading security measures taken by macOS High Sierra `

Or something

Comment (jondequinor) [reply .github/workflows/build.yml:161]: Perfect!

File: ert3/engine/_run.py (hunk 12 above)
Comment (jondequinor) [reply ert3/engine/_run.py:44]: Let's leave it as-is.

File: ert/storage/_storage.py (hunk 9 above)
Comment (jondequinor) [reply ert/storage/_storage.py:85]: It's fine that a transmitter subclass raises a NotImplementedError.

File: ert/storage/_storage.py (hunk 10 above)
Comment (DanSava) [reply ert/storage/_storage.py:114]: There is no implementation of the `RecordCollection` (former `EnsembleRecord`) that accounts for any transmitters currently. My PR just makes the previous code that was storing these RecordCollection to storage use transmitters, also getting these record collections from storage. And I guess because we are storing records in storage we use StorageTrasmitters.

There is no such thing as uniform shared-disk records, I am not sure there should be

File: ert/storage/_storage.py [hunk 15]
```diff
@@ -212,6 +279,82 @@ async def load_record(url: str, record_type: ert.data.RecordType) -> ert.data.Re
...
+    }
+
+    records_url = await get_records_url_async(workspace, experiment_name)
+    if experiment_name is not None:
+        ensemble_id = await _get_ensemble_id_async(workspace, experiment_name)
+        ensemble_size = await _get_ensemble_size(ensemble_id=ensemble_id)
+    else:
+        ensemble_size = record_coll.ensemble_size
+
+    if ensemble_size != record_coll.ensemble_size and record_coll.ensemble_size > 1:
```
[the remark this answers is not in the document]
Comment (DanSava) [reply ert/storage/_storage.py:349]: Got a bit lost in a rebase + plus a fix I think I will add it again

File: ert/storage/_storage.py [hunk 16]
```diff
@@ -99,28 +111,109 @@ async def _load_blob_record(self) -> ert.data.BlobRecord:
...
+
+    if ensemble_size is not None and len(transmitters) == 1:
+        return {iens: {record_name: transmitters[0]} for iens in range(ensemble_size)}
+    return {
+        iens: {record_name: transmitter}
+        for iens, transmitter in enumerate(transmitters)
+    }
+
+
+def get(url: str, headers: Dict[str, Any]) -> requests.Response:
```
[the remark this answers is not in the document]
Comment (DanSava) [reply ert/storage/_storage.py:149]: Sure, I will make it private, but I am sure there is a better way to structure this code, somebody should refactor this an try to make `httpx` or `aiohttp`

File: ert/storage/_storage.py [hunk 17]
```diff
@@ -212,6 +305,78 @@ async def load_record(url: str, record_type: ert.data.RecordType) -> ert.data.Re
     return ert.data.BlobRecord(data=content)
 
 
+async def get_record_metadata(record_url: str) -> Dict[Any, Any]:
+    headers = {
+        "Token": StorageInfo.token(),
+    }
+    # TODO once storage returns proper record metadata information add proper support for metadata
```
[the remark this answers is not in the document]
Comment (DanSava) [reply ert/storage/_storage.py:312]: Not really sure what issue you are referring to. There is no current issue in ert-storage to support in a different way metadata.  And I am not sure there will be. Moving forward we will use more the graphql API which will eliminate the need to stick some information in metadata

File: ert3/algorithms/_sensitivity.py [hunk 18]
```diff
@@ -118,7 +119,7 @@ def fast_sample(
 
 def fast_analyze(
     parameters: MutableMapping[str, Distribution],
-    model_output: RecordCollectionMap,
+    model_output: Dict[int, Dict[str, RecordTransmitter]],
```
Comment (xjules) [inline ert3/algorithms/_sensitivity.py:122]: Just wandering whether we shouldn't use `Mapping` instead of `Dict`; where `Mapping` implies immutability

Comment (DanSava) [reply ert3/algorithms/_sensitivity.py:122]: Not sure what the advantage of using Mapping is

File: ert3/engine/_export.py [hunk 19]
```diff
@@ -1,42 +1,94 @@
...
-        if not data:
-            data = [{"input": {}, "output": {}} for _ in ensemble_record.records]
+        elif record_source[0] == "resources":
+            record_mime = next(
+                input_.mime for input_ in step.input if input_.record == record_name
+            )
+            # DO NOT export blob records as inputs
+            if record_mime == "application/octet-stream":
+                continue
+            file_path = workspace / "resources" / record_source[1]
```
Comment (xjules) [inline ert3/engine/_export.py:52]: Use `os.path.join` ...

Comment (DanSava) [reply ert3/engine/_export.py:52]: Why? I thought the recommended way was to use pathlib.Path("some_path") / "something_else"?

File: ert3/algorithms/_sensitivity.py (hunk 18 above)
Comment (xjules) [reply ert3/algorithms/_sensitivity.py:122]: Not explicit, but it describes the parameter being immutable.

File: ert3/engine/_export.py (hunk 19 above)
Comment (xjules) [reply ert3/engine/_export.py:52]: Right, missed that `workspace` is a `Path` :) It's alright then...

**Comment by sondreso** [comment]
I'm ready to approve this, could we have a final rebase? 😄

**Review by jondequinor — CHANGES_REQUESTED** [review]
It appears this change breaks `ert3` (spe1 demo in the very least) on macOS (later versions). This means that before calling `ert3` you need to set `OBJC_DISABLE_INITIALIZE_FORK_SAFETY` every time. So you cannot fix this by extending `os.environ` _inside_ ERT.

But adding `OBJC_DISABLE_INITIALIZE_FORK_SAFETY` isn't a solution, it just prevents the underlying macOS SDK from crashing. It was crashing for a reason, which is undefined behavior caused _something_ calling`fork()` then accessing the macOS SDK before `exec()` (latter doesn't have to be called at all).

Using `forkserver` start method for `multiprocessing` should solve this, if that was the cause of the violation, but it doesn't appear that that is the case—or it still _might_ be. `forkserver` do invoke `fork()`, but it is supposed to be forking ERT in a very simple state (merely importing `main`, not executing it). However, we might be complicating that state somehow.

- So why was `forkserver` introduced? The default `spawn` method on macOS prevents the thread safety violation, since there's no "inheritance" of memory.
- If it isn't the `forkserver` that is violating thread safety, what is?

Either way, I don't think this should land as-is.

**Comment by jondequinor** [comment]
One hypothesis for why we need to disable fork safety is because the moment you introduce a http library into what `multiprocess` is bound to pickle, two things:
1. a http library will almost always do a macOS SDK call to obtain proxy information (we could try to set NO_PROXY=* and see if this mitigates it, of course not a solution)
2. you then get the `forkserver` `fork()` followed by a SDK call before `exec()`.

Last one is a bit speculative, but seems confirmed by the literature on this.

**Comment by jondequinor** [comment]
`NO_PROXY=*` mitigates this. Without knowing exactly what happens, this strongly hints towards the introduction of http libraries for objects relevant to code that passes the `fork()` barrier.

Paths forward:
 - a `forkserver` context should be created immediately after `if __name__ == "__main__":`, not as late as in the `PrefectEnsemble`.
 - abandoning the pickling strategy of Transmitters, migrating to using a well-known data format (yaml/json) instead.

**Comment by jondequinor** [comment]
The fork/exec issue can be triggered by introducing 
```

    async def _load_numerical_record(self) -> NumericalRecord:
        import requests
        requests.get("http://example.org")
```

into the `SharedStorageTransmitter`.

File: ert3/algorithms/_sensitivity.py [hunk 20]
```diff
@@ -131,22 +132,25 @@ def fast_analyze(
...
-
+    records = []
+    for transmitter_map in model_output.values():
+        if len(transmitter_map) > 1:
+            raise ValueError("Cannot analyze sensitivity with multiple outputs")
+        if len(transmitter_map) < 1:
+            raise ValueError("Cannot analyze sensitivity with no output")
+        for transmitter in transmitter_map.values():
+            records.append(
+                asyncio.get_event_loop().run_until_complete(transmitter.load())
```
Comment (jondequinor) [inline ert3/algorithms/_sensitivity.py:143]: batch this

File: ert_shared/ensemble_evaluator/evaluator.py [hunk 21]
```diff
@@ -77,7 +77,8 @@ async def _send_snapshot_update(self, snapshot_update_event):
             snapshot_update_event.to_dict(),
         )
         if event and self._clients:
-            await asyncio.wait([client.send(event) for client in self._clients])
+            # Fix deprecation warning
```
Comment (jondequinor) [inline ert_shared/ensemble_evaluator/evaluator.py:80]: needs fixin'

File: ert_shared/ensemble_evaluator/evaluator.py [hunk 22]
```diff
@@ -203,7 +204,10 @@ async def evaluator_server(self, done):
                 data_marshaller=cloudpickle.dumps,
             )
             if self._clients:
-                await asyncio.wait([client.send(message) for client in self._clients])
+                # Fixed deprecation warning
```
Comment (jondequinor) [inline ert_shared/ensemble_evaluator/evaluator.py:207]: needs fixin'

**Review by sondreso — APPROVED** [review]
:ship:
